lib/srv/regular/sshserver.go

```diff
@@ -1013,6 +1025,70 @@ func (s *Server) getServerResource() (types.Resource, error) {
 	return s.getServerInfo(), nil
 }
 
+func (s *Server) filterExistingNodes(instances *server.EC2Instances) (*server.EC2Instances, error) {
+	var filtered []*ec2.Instance
+	nodes, err := s.GetAccessPoint().GetNodes(s.ctx, s.getNamespace())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+outer:
+	for _, inst := range instances.Instances {
+		for _, node := range nodes {
+			if types.MatchLabels(node, map[string]string{
+				types.AWSAccountIDLabel:  instances.AccountID,
+				types.AWSInstanceIDLabel: aws.StringValue(inst.InstanceId),
+			}) {
+				continue outer
+			}
+		}
+		filtered = append(filtered, inst)
+	}
+	instances.Instances = filtered
+	return instances, nil
+}
+
+func (s *Server) handleInstances(instances *server.EC2Instances) error {
+	client, err := s.clients.GetAWSSSMClient(instances.Region)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	filteredInstances, err := s.filterExistingNodes(instances)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if len(filteredInstances.Instances) == 0 {
+		return trace.NotFound("all fetched nodes already registered")
+	}
+
+	installer := server.NewSSMInstaller(server.SSMInstallerConfig{Emitter: s})
+	req := server.SSMRunRequest{
```

Comment:
Can we trigger multiple times the SSM for the same machine? 
If for some reason the cluster/ssm is slow and the discovery runs multiple times before the SSM completes, can we end up with multiple SSM executions for the same machine?

SSM executions have costs, right?

Reply:
@lxea and I discussed this today, we're going to use auth server's semaphore to make sure only 1 agent runs SSM on the instance.